sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoDir.scala

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.hive.execution
+
+import java.io.File
+import java.util.Properties
+
+import scala.language.existentials
+
+import antlr.SemanticException
+import org.apache.hadoop.fs.{FileSystem, Path}
+import org.apache.hadoop.hive.ql.io.{HiveIgnoreKeyTextOutputFormat, RCFileOutputFormat}
+import org.apache.hadoop.hive.ql.io.orc.{OrcOutputFormat, OrcSerde}
+import org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat
+import org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe
+import org.apache.hadoop.hive.ql.plan.TableDesc
+import org.apache.hadoop.hive.serde.serdeConstants
+import org.apache.hadoop.hive.serde2.`lazy`.LazySimpleSerDe
+import org.apache.hadoop.io.Text
+import org.apache.hadoop.mapred.{FileOutputFormat, JobConf, SequenceFileOutputFormat, TextInputFormat}
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.catalog.CatalogStorageFormat
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.hive._
+import org.apache.spark.sql.hive.HiveShim.{ShimFileSinkDesc => FileSinkDesc}
+import org.apache.spark.util.{SerializableJobConf, Utils}
+
+case class InsertIntoDir(
+    path: String,
+    isLocal: Boolean,
+    fileFormat: String,
+    rowFormat: CatalogStorageFormat,
+    child: SparkPlan) extends SaveAsHiveFile {
+
+  @transient private val sessionState = sqlContext.sessionState.asInstanceOf[HiveSessionState]
+  def output: Seq[Attribute] = Seq.empty
+
+  protected[sql] lazy val sideEffectResult: Seq[InternalRow] = {
+    val hadoopConf = sessionState.newHadoopConf()
+
+    val properties = new Properties()
+
+    val Array(cols, types) = child.output.foldLeft(Array("", ""))((r, a) => {
+      r(0) = r(0) + a.name + ","
+      r(1) = r(1) + a.dataType.typeName + ":"
+      r
+    })
```

Comment:
Maybe it would be nicer if the additional closure is removed. This was discussed in (http://apache-spark-developers-list.1001551.n3.nabble.com/Question-about-Scala-style-explicit-typing-within-transformation-functions-and-anonymous-val-td17173.html)

```scala
val (cols, types) = child.output.foldLeft(("", "")) { case (r, a) =>
  r(0) = r(0) + a.name + ","
  r(1) = r(1) + a.dataType.typeName + ":"
  r
}
```